File: src/ui/theme.ts
/**
 * Theme System for UI Components
 * @module ui/theme
 */

/**
 * Theme color definitions
 */
export interface ThemeColors {
  // Base
  background: string;
  foreground: string;
  text: string;
  textDim: string;

  // Accent
  primary: string;
  secondary: string;
  accent: string;

  // Status
  success: string;
  warning: string;
  error: string;
  info: string;

  // UI Elements
  border: string;
  borderActive: string;
  headerBg: string;
  headerFg: string;

  // Keybinds
  keybindGlobal: string;
  keybindContext: string;

  // Window Switcher
  windowCurrent: string;
  windowAdjacent: string;

  // Wizard
  wizardSystem: string;
  wizardUser: string;
  wizardOptionSelected: string;
  wizardOptionNormal: string;

  // Status Bar
  statusInfo: string;
  statusSuccess: string;
  statusWarning: string;
  statusError: string;

  // Escape Indicator
  escapeIndicatorActive: string;
  escapeIndicatorInactive: string;

  // Screensaver
  screensaverBg: string;
  screensaverFg: string;
}

/**
 * Theme class for managing UI colors
 */
export class Theme {
  private constructor(private colors: ThemeColors) {}

  /**
   * Get the default dark theme (Catppuccin-inspired)
   */
  static defaultDark(): Theme {
    return new Theme({
      // Base
      background: '#1e1e2e',
      foreground: '#cdd6f4',
      text: '#cdd6f4',
      textDim: '#6c7086',

      // Accent
      primary: '#89b4fa',
      secondary: '#94e2d5',
      accent: '#f38ba8',

      // Status
      success: '#a6e3a1',
      warning: '#f9e2af',
      error: '#f38ba8',
      info: '#89dceb',

      // UI Elements
      border: '#45475a',
      borderActive: '#89b4fa',
      headerBg: '#313244',
      headerFg: '#cdd6f4',

      // Keybinds
      keybindGlobal: '#89b4fa',
      keybindContext: '#f9e2af',

      // Window Switcher
      windowCurrent: '#a6e3a1',
      windowAdjacent: '#6c7086',

      // Wizard
      wizardSystem: '#89dceb',
      wizardUser: '#f5c2e7',
      wizardOptionSelected: '#a6e3a1',
      wizardOptionNormal: '#cdd6f4',

      // Status Bar
      statusInfo: '#89dceb',
      statusSuccess: '#a6e3a1',
      statusWarning: '#f9e2af',
      statusError: '#f38ba8',

      // Escape Indicator
      escapeIndicatorActive: '#f38ba8',
      escapeIndicatorInactive: '#45475a',

      // Screensaver
      screensaverBg: '#000000',
      screensaverFg: '#a6e3a1',
    });
  }

  /**
   * Get a color by key
   * @param key Color key
   * @returns Color value (hex string)
   */
  get<K extends keyof ThemeColors>(key: K): string {
    return this.colors[key];
  }

  /**
   * Get all colors
   * @returns All theme colors
   */
  getAll(): ThemeColors {
    return { ...this.colors };
  }
}

/**
 * ANSI color codes for terminal output
 */
export const ANSI = {
  reset: '\x1b[0m',
  bold: '\x1b[1m',
  dim: '\x1b[2m',
  italic: '\x1b[3m',
  underline: '\x1b[4m',
  blink: '\x1b[5m',
  reverse: '\x1b[7m',
  hidden: '\x1b[8m',

  // Foreground colors
  black: '\x1b[30m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  white: '\x1b[37m',

  // Background colors
  bgBlack: '\x1b[40m',
  bgRed: '\x1b[41m',
  bgGreen: '\x1b[42m',
  bgYellow: '\x1b[43m',
  bgBlue: '\x1b[44m',
  bgMagenta: '\x1b[45m',
  bgCyan: '\x1b[46m',
  bgWhite: '\x1b[47m',

  // Bright foreground colors
  brightBlack: '\x1b[90m',
  brightRed: '\x1b[91m',
  brightGreen: '\x1b[92m',
  brightYellow: '\x1b[93m',
  brightBlue: '\x1b[94m',
  brightMagenta: '\x1b[95m',
  brightCyan: '\x1b[96m',
  brightWhite: '\x1b[97m',
} as const;

/**
 * Convert hex color to ANSI 256 color code
 * @param hex Hex color string
 * @returns ANSI escape sequence for the color
 */
export function hexToAnsi256(hex: string): string {
  // Remove # if present
  hex = hex.replace('#', '');

  const r = parseInt(hex.substring(0, 2), 16);
  const g = parseInt(hex.substring(2, 4), 16);
  const b = parseInt(hex.substring(4, 6), 16);

  // Convert to 256 color
  const code = 16 + 36 * Math.round((r / 255) * 5) + 6 * Math.round((g / 255) * 5) + Math.round((b / 255) * 5);

  return `\x1b[38;5;${code}m`;
}

/**
 * Convert hex color to ANSI 256 background color code
 * @param hex Hex color string
 * @returns ANSI escape sequence for the background color
 */
export function hexToAnsi256Bg(hex: string): string {
  // Remove # if present
  hex = hex.replace('#', '');

  const r = parseInt(hex.substring(0, 2), 16);
  const g = parseInt(hex.substring(2, 4), 16);
  const b = parseInt(hex.substring(4, 6), 16);

  // Convert to 256 color
  const code = 16 + 36 * Math.round((r / 255) * 5) + 6 * Math.round((g / 255) * 5) + Math.round((b / 255) * 5);

  return `\x1b[48;5;${code}m`;
}

/**
 * Apply color to text using ANSI codes
 * @param text Text to colorize
 * @param fg Foreground hex color
 * @param bg Optional background hex color
 * @returns Colorized text with ANSI codes
 */
export function colorize(text: string, fg: string, bg?: string): string {
  let result = hexToAnsi256(fg);
  if (bg) {
    result += hexToAnsi256Bg(bg);
  }
  return result + text + ANSI.reset;
}

File: src/ui/theme.rs
use ratatui::style::{Color, Modifier, Style};

/// Theme struct containing all colors and styles for the application
/// Based on v2.1 spec section 13
#[derive(Debug, Clone)]
pub struct Theme {
    // Base colors
    pub background: Color,
    pub foreground: Color,
    pub text: Color,
    pub text_dim: Color,
    
    // Accent colors
    pub primary: Color,
    pub secondary: Color,
    pub accent: Color,
    
    // Status colors
    pub success: Color,
    pub warning: Color,
    pub error: Color,
    pub info: Color,
    
    // UI elements
    pub border: Color,
    pub border_active: Color,
    pub header_bg: Color,
    pub header_fg: Color,
    
    // Keybind bar
    pub keybind_global: Color,
    pub keybind_context: Color,
    
    // Window switcher
    pub window_current: Color,
    pub window_adjacent: Color,
    
    // Wizard
    pub wizard_system: Color,
    pub wizard_user: Color,
    pub wizard_option_selected: Color,
    pub wizard_option_normal: Color,
    
    // Status bar
    pub status_info: Style,
    pub status_success: Style,
    pub status_warning: Style,
    pub status_error: Style,
    
    // Escape indicator
    pub escape_indicator_active: Color,
    pub escape_indicator_inactive: Color,
    
    // Screensaver
    pub screensaver_bg: Color,
    pub screensaver_fg: Color,
}

impl Theme {
    /// Default dark theme
    pub fn default_dark() -> Self {
        Self {
            background: Color::Black,
            foreground: Color::White,
            text: Color::Rgb(220, 220, 220),
            text_dim: Color::Rgb(128, 128, 128),
            
            primary: Color::Cyan,
            secondary: Color::Blue,
            accent: Color::Yellow,
            
            success: Color::Green,
            warning: Color::Yellow,
            error: Color::Red,
            info: Color::Cyan,
            
            border: Color::DarkGray,
            border_active: Color::Cyan,
            header_bg: Color::Rgb(30, 30, 30),
            header_fg: Color::White,
            
            keybind_global: Color::Yellow,
            keybind_context: Color::Rgb(180, 180, 180),
            
            window_current: Color::Cyan,
            window_adjacent: Color::DarkGray,
            
            wizard_system: Color::Cyan,
            wizard_user: Color::Green,
            wizard_option_selected: Color::Yellow,
            wizard_option_normal: Color::White,
            
            status_info: Style::default().fg(Color::Cyan),
            status_success: Style::default().fg(Color::Green).add_modifier(Modifier::BOLD),
            status_warning: Style::default().fg(Color::Yellow).add_modifier(Modifier::BOLD),
            status_error: Style::default().fg(Color::Red).add_modifier(Modifier::BOLD),
            
            escape_indicator_active: Color::Yellow,
            escape_indicator_inactive: Color::DarkGray,
            
            screensaver_bg: Color::Black,
            screensaver_fg: Color::Green, // Matrix-style
        }
    }
}

impl Default for Theme {
    fn default() -> Self {
        Self::default_dark()
    }
}
